File: src/Timetable.ts
export interface Timetable {
  timetable: Lesson[];
  target: string;
  type: TimetableTypeEnum;
}

export interface Lesson {
  day: DayEnum;
  week: WeekEnum;
  time: LessonTime;
  subject: string;
  type: string;
  place: string;
  teacher: string | null;
  groups: string[] | null;
}

export enum DayEnum {
  Monday = "1",
  Tuesday = "2",
  Wednesday = "3",
  Thursday = "4",
  Friday = "5",
  Saturday = "6",
  Sunday = "0",
}

export enum TimetableTypeEnum {
  Teacher = "teacher",
  Group = "group",
}

export enum WeekEnum {
  Uneven = "1",
  Even = "2",
}

export type LessonTime =
  | "8:30-10:05"
  | "10:15-11:50"
  | "12:00-13:35"
  | "14:10-15:45"
  | "15:55-17:30"
  | "17:40-19:15"
  | "19:25-21:00";

export type DayOfWeek =
  | "Воскресенье"
  | "Понедельник"
  | "Вторник"
  | "Среда"
  | "Четверг"
  | "Пятница"
  | "Суббота";

export const DayEnumFromDayNumber = (day: number): DayEnum => {
  switch (day) {
    case 1: {
      return DayEnum.Monday;
    }
    case 2: {
      return DayEnum.Tuesday;
    }
    case 3: {
      return DayEnum.Wednesday;
    }
    case 4: {
      return DayEnum.Thursday;
    }
    case 5: {
      return DayEnum.Friday;
    }
    case 6: {
      return DayEnum.Saturday;
    }
    case 0: {
      return DayEnum.Sunday;
    }
    default: {
      throw new Error("Нет такого дня недели");
    }
  }
};

export const DayOfWeekFromNumber = (day: DayEnum): DayOfWeek => {
  switch (day) {
    case DayEnum.Monday: {
      return "Понедельник";
    }
    case DayEnum.Tuesday: {
      return "Вторник";
    }
    case DayEnum.Wednesday: {
      return "Среда";
    }
    case DayEnum.Thursday: {
      return "Четверг";
    }
    case DayEnum.Friday: {
      return "Пятница";
    }
    case DayEnum.Saturday: {
      return "Суббота";
    }
    case DayEnum.Sunday: {
      return "Воскресенье";
    }
  }
};

export const GetWeekNum = (date: Date): WeekEnum => {
  let day = date.getDate();
  let month = date.getMonth();
  let year = date.getFullYear();
  month++;
  let a = Math.floor((14 - month) / 12);
  let y = year + 4800 - a;
  let m = month + 12 * a - 3;
  let J =
    day +
    Math.floor((153 * m + 2) / 5) + 365 * y +
    Math.floor(y / 4) -
    Math.floor(y / 100) +
    Math.floor(y / 400) - 32045;
  let d4 = (((J + 31741 - (J % 7)) % 146097) % 36524) % 1461;
  let L = Math.floor(d4 / 1460);
  let d1 = ((d4 - L) % 365) + L;
  let week = Math.floor(d1 / 7) + 1;
  return week % 2 === 0 ? WeekEnum.Uneven : WeekEnum.Even;
};


export const FetchTimetable = async (target: string) => {
  const res = await fetch(
    "https://edu.sfu-kras.ru/api/timetable/get&target=" + encodeURIComponent(target)
  );
  return await res.json() as Timetable;
}

File: src/shared/Timetable.ml
module Option = Belt.Option
module Result = Belt.Result
module Json = Js.Json
module Dict = Js.Dict

let decodeStringExn obj key =
  obj |. Dict.get key |. Option.flatMap Json.decodeString |. Option.getExn

module Day = struct
  type t =
    | Monday
    | Tuesday
    | Wednesday
    | Thursday
    | Friday
    | Saturday
    | Sunday
  [@@genType]

  let toNumber day =
    match day with
    | Monday -> 1
    | Tuesday -> 2
    | Wednesday -> 3
    | Thursday -> 4
    | Friday -> 5
    | Saturday -> 6
    | Sunday -> 7
    [@@genType]

  let fromNumber number =
    match number with
    | 1 -> Monday
    | 2 -> Tuesday
    | 3 -> Wednesday
    | 4 -> Thursday
    | 5 -> Friday
    | 6 -> Saturday
    | 7 | 0 -> Sunday
    | _ -> failwith "No day of week"
    [@@genType]

  let decodeExn json =
    Json.decodeString json |> Option.getExn |> Belt.Int.fromString
    |> Option.getExn |> fromNumber

  let toRu day =
    match day with
    | Monday -> {j|Понедельник|j}
    | Tuesday -> {j|Вторник|j}
    | Wednesday -> {j|Среда|j}
    | Thursday -> {j|Четверг|j}
    | Friday -> {j|Пятница|j}
    | Saturday -> {j|Суббота|j}
    | Sunday -> {j|Воскресенье|j}
    [@@genType]

  type en =
    [ `Monday
    | `Tuesday
    | `Wednesday
    | `Thursday
    | `Friday
    | `Saturday
    | `Sunday ]
  [@@genType]

  let toEn day : en =
    match day with
    | Monday -> `Monday
    | Tuesday -> `Tuesday
    | Wednesday -> `Wednesday
    | Thursday -> `Thursday
    | Friday -> `Friday
    | Saturday -> `Saturday
    | Sunday -> `Sunday
    [@@genType]

  let all =
    [| Monday; Tuesday; Wednesday; Thursday; Friday; Saturday; Sunday |]
    [@@genType]
end

module Week = struct
  type t = Uneven | Even [@@genType]

  let fromNumber n =
    match n with 1 -> Uneven | 2 -> Even | _ -> failwith ""
    [@@genType]

  let decodeExn json =
    Json.decodeString json |> Option.getExn |> Belt.Int.fromString
    |> Option.getExn |> fromNumber

  let fromDate (date : Js.Date.t) =
    let day = date |. Js.Date.getDate |. Belt.Int.fromFloat in
    let month = date |. Js.Date.getMonth |. Belt.Int.fromFloat in
    let year = date |. Js.Date.getFullYear |. Belt.Int.fromFloat in
    let month = month + 1 in
    let a = Js.Math.floor ((14. -. (month |. Belt.Int.toFloat)) /. 12.) in
    let y = year + 4800 - a in
    let m = month + (12 * a) - 3 in
    let j =
      day
      + Js.Math.floor (((153. *. (m |. Belt.Int.toFloat)) +. 2.) /. 5.)
      + (365 * y)
      + Js.Math.floor ((y |. Belt.Int.toFloat) /. 4.)
      - Js.Math.floor ((y |. Belt.Int.toFloat) /. 100.)
      + Js.Math.floor ((y |. Belt.Int.toFloat) /. 400.)
      - 32045
    in
    let d4 = (j + 31741 - (j mod 7)) mod 146097 mod 36524 mod 1461 in
    let l = Js.Math.floor ((d4 |. Belt.Int.toFloat) /. 1460.) in
    let d1 = ((d4 - l) mod 365) + l in
    let week = Js.Math.floor ((d1 |. Belt.Int.toFloat) /. 7.) + 1 in
    if week mod 2 == 0 then Even else Uneven
    [@@genType]
end

module Lesson = struct
  type t = {
    day : Day.t;
    week : Week.t;
    time : LessonTime.t;
    subject : string;
    type_ : string;
    teacher : string option;
    groups : string array option;
    place : string;
    building : string;
    room : string;
    sync : string;
  }
  [@@genType]

  let decodeExn json =
    try
      let obj = json |> Json.decodeObject |> Option.getExn in
      let day =
        obj |. Dict.get "day" |. Option.map Day.decodeExn |> Option.getExn
      in
      let week =
        obj |. Dict.get "week" |. Option.map Week.decodeExn |> Option.getExn
      in
      let teacher =
        obj |. Dict.get "teacher" |. Option.flatMap Json.decodeString
      in
      let groups =
        obj |. Dict.get "groups"
        |. Option.flatMap Json.decodeArray
        |. Option.map (Array.map Json.decodeString)
        |. Option.map (Array.map Option.getExn)
      in

      {
        day;
        week;
        time =
          obj |. Dict.get "time"
          |. Option.map LessonTime.decode
          |> Option.getExn;
        subject =
          obj |. Dict.get "subject"
          |. Option.flatMap Json.decodeString
          |> Option.getExn;
        type_ =
          obj |. Dict.get "type"
          |. Option.flatMap Json.decodeString
          |> Option.getExn;
        teacher;
        groups;
        place =
          obj |. Dict.get "place"
          |. Option.flatMap Json.decodeString
          |> Option.getExn;
        building =
          obj |. Dict.get "building"
          |. Option.flatMap Json.decodeString
          |> Option.getExn;
        room =
          obj |. Dict.get "room"
          |. Option.flatMap Json.decodeString
          |> Option.getExn;
        sync =
          obj |. Dict.get "sync"
          |. Option.flatMap Json.decodeString
          |> Option.getExn;
      }
    with error ->
      Js.Console.error error;
      failwith @@ "Failed to decode Lesson"
end

module Timetable = struct
  type t = {
    timetable : Lesson.t array;
    target : string;
    type_ : [ `teacher | `group ];
    institute : string option;
  }
  [@@genType]

  let decode json =
    try
      let obj = json |. Json.decodeObject |> Option.getExn in
      let timetable =
        obj |. Dict.get "timetable"
        |. Option.flatMap Json.decodeArray
        |. Option.map (Array.map Lesson.decodeExn)
        |. Option.getExn
      in
      Ok
        {
          timetable;
          target = obj |. decodeStringExn "target";
          type_ =
            obj |. Dict.get "type"
            |. Option.flatMap Json.decodeString
            |. Option.getExn |. Obj.magic;
          institute =
            obj |. Dict.get "institute" |. Option.flatMap Json.decodeString;
        }
    with error ->
      Js.Console.error error;
      Error "Failed to decode Timetable"

  let fetch target : _ Js.Promise.t =
    let open Compat in
    Fetch.fetch
      ( "https://edu.sfu-kras.ru/api/timetable/get&target="
      ^ Js.Global.encodeURIComponent target )
    |. Promise.then_ (fun response -> Fetch.Response.json response)
    |. Promise.thenResolve (fun json -> decode json)
    [@@genType]

  let fetchOpt target : _ Js.Promise.t =
    fetch target
    |. Promise.thenResolve (function Ok a -> Some a | Error _ -> None)
    [@@genType]

  let fetchExn target : _ Js.Promise.t =
    fetch target
    |. Promise.thenResolve (function
         | Ok a -> a
         | Error error -> failwith error)
    [@@genType]
end
